relay-server/src/actors/store.rs

```diff
@@ -199,51 +199,117 @@ impl StoreService {
             }
         }
 
+        let remote_addr = envelope.meta().client_addr().map(|addr| addr.to_string());
+
+        let kafkamessages = self.extract_kafka_messages(
+            event_item,
+            event_id,
+            scoping,
+            start_time,
+            remote_addr,
+            attachments,
+        );
+
+        for message in kafkamessages {
+            if let Ok(()) = self.produce(topic, scoping.organization_id, &message) {
+                // If message was succesfully sent, make the right log message.
+                match message {
+                    KafkaMessage::Attachment(_) => {
+                        metric!(
+                            counter(RelayCounters::ProcessingMessageProduced) += 1,
+                            event_type = "attachment"
+                        );
+                    }
+                    KafkaMessage::Event(_) => {
+                        if let Some(event_item) = event_item {
+                            metric!(
+                                counter(RelayCounters::ProcessingMessageProduced) += 1,
+                                event_type = &event_item.ty().to_string()
+                            );
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn extract_kafka_messages(
+        &self,
+        event_item: Option<&Item>,
+        event_id: Option<EventId>,
+        scoping: Scoping,
+        start_time: Instant,
+        remote_addr: Option<String>,
+        mut attachments: Vec<ChunkedAttachment>,
+    ) -> Vec<KafkaMessage> {
```

Comment:
What would be the benefit of returning an iterator instead? Or do you mean that on `fn extract_kafka_attachment` return an iterator instead of mutating the Vec<KafkaMessage>?

Reply:
Iterators are a [zero-cost abstraction](https://doc.rust-lang.org/book/ch13-04-performance.html#:~:text=Iterators%20are%20one%20of%20Rust's,of%20C%2B%2B%E2%80%9D%20(2012)%3A), which means its runtime cost is just as low as having a simple for-loop. Vectors on the other hand require memory allocations, and cause data to being copied around.